Item Components Update

jeantkg opened the pull request:
I have added all the block components necessary for custom blocks, I also remove the use of the Model and Material file as we can use block components instead.

TwistedAsylumMC commented on src/block/BlockComponents.php:
```diff
@@ -0,0 +1,29 @@
+interface BlockComponents {
+
+    /**
+     * Add component adds a component to the item that can be returned in the getComponents() method to be sent over
```
block not item

jeantkg replied:
that text was a copy and paste from the Item Component file

TwistedAsylumMC commented on src/block/BlockComponents.php:
```diff
@@ -0,0 +1,29 @@
+    public function addComponent(BlockComponent $component): void;
+
+    /**
+     * Returns if the item has the component with the provided name.
```
block not item

TwistedAsylumMC commented on src/block/BlockComponents.php:
```diff
@@ -0,0 +1,29 @@
+    public function hasComponent(string $name): bool;
+
+    /**
+     * @return BlockComponents
```
[]BlockComponent ?

TwistedAsylumMC commented on src/block/BlockComponentsTrait.php:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    /**
+     * @return array
```
BlockComponent[]

TwistedAsylumMC commented on src/block/CustomiesBlockFactory.php:
```diff
@@ -93,21 +94,7 @@ public function registerBlock(Closure $blockFunc, string $identifier, ?Model $mo
 		$this->customBlocks[$identifier] = $block;
 
 		$propertiesTag = CompoundTag::create();
-		$components = CompoundTag::create()
-			->setTag("minecraft:light_emission", CompoundTag::create()
-				->setByte("emission", $block->getLightLevel()))
-			->setTag("minecraft:light_dampening", CompoundTag::create()
-				->setByte("lightLevel", $block->getLightFilter()))
-			->setTag("minecraft:destructible_by_mining", CompoundTag::create()
-				->setFloat("value", $block->getBreakInfo()->getHardness()))
-			->setTag("minecraft:friction", CompoundTag::create()
-				->setFloat("value", 1 - $block->getFrictionFactor()));
-
-		if($model !== null) {
-			foreach($model->toNBT() as $tagName => $tag){
-				$components->setTag($tagName, $tag);
-			}
-		}
```
Should there be any default components?

jeantkg replied:
I asked a question in discord on this

TwistedAsylumMC commented on src/block/component/BlockComponent.php:
```diff
@@ -0,0 +1,11 @@
+
+    public function getValue(): mixed;
+
+    // public function isProperty(): bool;
```
what is this

TwistedAsylumMC commented on src/block/component/CraftingTableComponent.php:
```diff
@@ -0,0 +1,27 @@
+    private string $tableName;
+    private array $craftingTags;
+
+    public function __construct(string $tableName, array $craftingTags) {
+		$this->tableName = $tableName;
+        $this->craftingTags = $craftingTags;
+	}
```
Formatting

TwistedAsylumMC commented on src/block/component/CraftingTableComponent.php:
```diff
@@ -0,0 +1,27 @@
+class CraftingTableComponent implements BlockComponent {
+
+    private string $tableName;
+    private array $craftingTags;
```
Annotate array type

TwistedAsylumMC commented on src/block/component/CraftingTableComponent.php:
```diff
@@ -0,0 +1,27 @@
+    public function getValue(): CompoundTag {
+        return CompoundTag::create()
+            ->setString("table_name", $this->tableName)
+            ->setTag("crafting_tags", new ListTag([$this->craftingTags]));
```
Is this meant to be [$this->craftingTags], will that not be [[tag, tag]]?

jeantkg replied:
yes well that is what I understood from [this](https://learn.microsoft.com/en-us/minecraft/creator/reference/content/blockreference/examples/blockcomponents/minecraftblock_crafting_table?view=minecraft-bedrock-stable)

TwistedAsylumMC commented on src/block/component/DestructibleByExplosionComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private float $explosionResistance;
+
+    public function __construct(float $explosionResistance = 0.0) {
+		$this->explosionResistance = $explosionResistance;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/DestructibleByMiningComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private float $secondsToDestroy;
+
+    public function __construct(float $secondsToDestroy = 0.0) {
+		$this->secondsToDestroy = $secondsToDestroy;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/DisplayNameComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private string $displayName;
+
+    public function __construct(string $displayName) {
+		$this->displayName = $displayName;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/FlammableComponent.php:
```diff
@@ -0,0 +1,27 @@
+
+    private int $destroyChanceModifier;
+
+    public function __construct(int $catchChanceModifier = 5, int $destroyChanceModifier = 20) {
+		$this->catchChanceModifier = $catchChanceModifier;
+        $this->destroyChanceModifier = $destroyChanceModifier;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/FlammableComponent.php:
```diff
@@ -0,0 +1,27 @@
+class FlammableComponent implements BlockComponent {
+
+    private int $catchChanceModifier;
+
```
remove line

TwistedAsylumMC commented on src/block/component/FrictionComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private float $decimal;
+
+    public function __construct(float $decimal = 0.4) {
+		$this->decimal = $decimal;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/FrictionComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+class FrictionComponent implements BlockComponent {
+
+    private float $decimal;
```
why "decimal"?

TwistedAsylumMC commented on src/block/component/GeometryComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private string $geometry;
+
+    public function __construct(string $geometry) {
+		$this->geometry = $geometry;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/GeometryComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    public function getValue(): CompoundTag {
+        return CompoundTag::create()
+			->setString("identifier", $this->geometry);
```
formatting

TwistedAsylumMC commented on src/block/component/LightDampeningComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private int $integer;
+
+    public function __construct(int $integer = 15) {
+		$this->integer = $integer;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/LightDampeningComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+class LightDampeningComponent implements BlockComponent {
+
+    private int $integer;
```
could have a better name

TwistedAsylumMC commented on src/block/component/LightEmissionComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private int $integer;
+
+    public function __construct(int $integer = 0) {
+		$this->integer = $integer;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/LootComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private string $pathString;
+
+    public function __construct(string $pathString) {
+		$this->pathString = $pathString;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/MapColorComponent.php:
```diff
@@ -0,0 +1,23 @@
+
+    private string $hexString;
+
+    public function __construct(string $hexString) {
+		$this->hexString = $hexString;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/MaterialInstancesComponent.php:
```diff
@@ -0,0 +1,58 @@
+
+class MaterialInstancesComponent implements BlockComponent {
+
+    public const TARGET_ALL = "*";
+	public const TARGET_SIDES = "sides";
+	public const TARGET_UP = "up";
+	public const TARGET_DOWN = "down";
+	public const TARGET_NORTH = "north";
+	public const TARGET_EAST = "east";
+	public const TARGET_SOUTH = "south";
+	public const TARGET_WEST = "west";
+
+	public const RENDER_METHOD_ALPHA_TEST = "alpha_test";
+	public const RENDER_METHOD_BLEND = "blend";
+	public const RENDER_METHOD_OPAQUE = "opaque";
```
formatting

TwistedAsylumMC commented on src/block/component/MaterialInstancesComponent.php:
```diff
@@ -0,0 +1,58 @@
+    private bool $faceDimming;
+    private string $renderMethod;
+
+    public function __construct(
+        string $target, 
+        string $texture, 
+        string $renderMethod = self::RENDER_METHOD_OPAQUE, 
+        bool $ambientOcclusion = true, 
+        bool $faceDimming = true
+    ) {
+		$this->target = $target;
+        $this->texture = $texture;
+        $this->ambientOcclusion = $ambientOcclusion;
+        $this->faceDimming = $faceDimming;
+        $this->renderMethod = $renderMethod;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/SelectionBoxComponent.php:
```diff
@@ -0,0 +1,38 @@
+    private Vector3 $origin;
+    private Vector3 $size;
+
+    public function __construct(Vector3 $origin = new Vector3(-8.0, 0.0, -8.0), Vector3 $size = new Vector3(16.0, 16.0, 16.0)) {
+		$this->origin = $origin;
+        $this->size = $size;
+	}
```
formatting

TwistedAsylumMC commented on src/block/component/TransformationComponent.php:
```diff
@@ -0,0 +1,64 @@
+    private Vector3 $rotation;
+    private Vector3 $rotationPivot;
+
+    public function __construct(
+        Vector3 $translation = new Vector3(0, 0, 0), 
+        Vector3 $scale = new Vector3(1, 1, 1), 
+        Vector3 $scalePivot = new Vector3(0, 0, 0), 
+        Vector3 $rotation = new Vector3(0, 0, 0), 
+        Vector3 $rotationPivot = new Vector3(0, 0, 0)
+    ) {
+		$this->translation = $translation;
+        $this->scale = $scale;
+        $this->scalePivot = $scalePivot;
+        $this->rotation = $rotation;
+        $this->rotationPivot = $rotationPivot;
+	}
```
formatting

TwistedAsylumMC commented on composer.json:
```diff
@@ -0,0 +1,17 @@
+{
+	"name": "customiesdevs/customies",
+	"type": "library",
+	"minimum-stability": "beta",
+	"prefer-stable": true,
+	"autoload": {
+		"classmap": ["src"]
+	},
+	"scripts": {
+		"build": "php -dphar.readonly=0 vendor/bin/pharynx -i=. -c -p=Timer.phar"
+	},
+	"config": {
+		"allow-plugins": {
+			"phpstan/extension-installer": true
+		}
+	}
+}
```
Is this needed?

jeantkg replied:
nope, this wasn't supposed to be pushed

TwistedAsylumMC commented on src/block/BlockComponentsTrait.php:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	/**
+	 * @return BlockComponent
```
also an array

jeantkg replied:
check CustomiesBlockFactory line 165 - 169

TwistedAsylumMC commented on src/block/BlockComponentsTrait.php:
```diff
@@ -0,0 +1,43 @@
+	/**
+	 * Initializes the block with default components that are required for the block to function correctly.
+	 */
+	protected function initComponent(): void {
```
The values of these can be inferred from the block if it implements certain PM classes, similar to items

jeantkg replied:
yeah, will work on this tomorrow hopefully

jeantkg commented:
I'm going to also update Item Components in this Pull Request. Also is there any reason for isProperty() anymore? Is there any docs on item_properties?

Nya-Enzo commented:
the item_properties are mainly protocol related, and you most likely won't see them in any docs since addons don't use it like that, it's just mojang requires it to be like this in the packet data

jeantkg commented:
Alright, this PR is done, I added new Item Components and updated the others.
These are the components I didn't add:
- EntityPlacerComponent
- ItemTagsComponent & TagsComponent
- RepairableComponent
- StorageComponent
These components need their own PR as they are very complicated.

TwistedAsylumMC commented on src/item/component/DamageComponent.php:
```diff
@@ -0,0 +1,29 @@
+
+	/**
+	 * Determines how much extra damage an item does on attack. Note that this must be a positive value.
+	 * @param int $damage Should be a Intger above `0`
```
If it should be >0, why is the default 0? I'd argue a default is not necessary here

jeantkg replied:
yes your right there shouldn't be a default value in the first place, don't really know how that got added

TwistedAsylumMC commented on src/item/component/DurabilityComponent.php:
```diff
@@ -17,6 +27,10 @@ public function getName(): string {
 
 	public function getValue(): array {
 		return [
+			"damage_chance" => [
+				"min" => $this->minDamageChance,
+				"max" => $this->maxDamageChance
+			],
```
Is it necessary adding these if pocketmine doesn't support it?

jeantkg replied:
yes for future support and they won't really do harm as I added default values for them

TwistedAsylumMC commented on src/item/component/HoverTextColorComponent.php:
```diff
@@ -0,0 +1,30 @@
+	 * @param string $hoverTextColor Specifies the color of the item's hover text
+	 * @link [List of Color Code](https://minecraft.wiki/w/Formatting_codes#Color_codes)
+	 */
+	public function __construct(string $hoverTextColor = "§s") {
```
Why is this the default?

jeantkg replied:
Oh I think I added that for testing purposes, there isn't a default so I'll remove it

TwistedAsylumMC commented on src/item/component/UseDurationComponent.php:
```diff
@@ -7,7 +7,11 @@ final class UseDurationComponent implements ItemComponent {
-	public function __construct(int $duration) {
+	/**
+	 * How long the item takes to use in seconds.
+	 * @param int $duration Default is set to `20` (`1` means `20` seconds) so (`1.6` means `32` seconds)
+	 */
+	public function __construct(int $duration = 20) {
```
If duration=1 means 20 seconds, does this default not mean it'll take 400 seconds?

jeantkg replied:
I should switch the wording here as it is supposed to be 20 means 1 second

TwistedAsylumMC commented on src/item/component/ProjectileComponent.php:
```diff
@@ -17,6 +26,7 @@ public function getName(): string {
 
 	public function getValue(): array {
 		return [
+			"minimum_critical_power" => $this->minimumCriticalPower,
```
Is this needed to be sent to the client if it's not supported by PM?

jeantkg replied:
again future support and there is already a default value for this

TwistedAsylumMC reviewed (changes requested):
Just a few small comments, have you also had the chance to test all the components like rarity, bundle etc?

jeantkg commented on src/item/component/DamageAbsorptionComponent.php:
```diff
@@ -0,0 +1,32 @@
+
+final class DamageAbsorptionComponent implements ItemComponent {
+
+	private array $absorbableCauses;
```
Oh I missed one

jeantkg replied:
So for this component, I got it from the [bedrock devs website ](https://wiki.bedrock.dev/items/item-components#damage-absorption) and there is no mention of it in the minecraft creator docs. So I have no idea of what all the absorbable causes are.

HydroGames-dev replied:
> absorbable

btw there a default value for this check my britt-fork

jeantkg commented:
> Just a few small comments, have you also had the chance to test all the components like rarity, bundle etc?

Not all components have been tested but rarity and bundles have been tested and I can show some screenshots of them, HydroGames / Britty created a backpack using the bundle component.
I should get you a list of components that haven't been tested.

TwistedAsylumMC commented:
Thank you for this change, I saw that you're also interested in updating the wiki so that would be greatly appreciated!
